### PR TITLE
BUG: group_shift_indexer checks for null group keys

### LINKED ISSUE #13813: Groupby and shift causing coredump
Hi, 

I have attempted to reduce this to the smallest example that exhibits this issue - rather than a useful example. The problem is that the operation causes python to core dump.

In the original case in which I discovered this the core dump would only occur sometimes (and when I put it in a loop it would occur on different iterations). This code seems to core dump on the third iteration every time I have run it.

Code example:

```
import os
import pandas as pd

df = pd.read_csv(os.path.join(os.getcwd(), 'error_report.txt'), sep='\t')

for i in range(0, 1000):
    print "Pre shift {}".format(i)
    df['shift_F'] = df.groupby(['B', 'C'])['F'].shift(-1)
    print "Post shift {}".format(i)

```
With the attached data file (tab separated) - code assumes in the same directory:
[error_report.txt](https://github.com/pydata/pandas/files/385604/error_report.txt)

This the output I get:

```
python pandas_test.py 
Pre shift 0
Post shift 0
Pre shift 1
Post shift 1
Pre shift 2
Segmentation fault (core dumped)

```
If I modify the code to use apply and then add the shifted column inside the apply function then there is no error. Similarly if I use .shift(0) I do not get the error.

Version info:

```
>>> pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8

pandas: 0.18.1
nose: 1.3.1
pip: 8.1.2
setuptools: 20.2.2
Cython: None
numpy: 1.11.1
scipy: 0.13.3
statsmodels: 0.5.0
xarray: None
IPython: 1.2.1
sphinx: None
patsy: 0.2.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: None
tables: 3.1.1
numexpr: 2.6.1
matplotlib: 1.3.1
openpyxl: 1.7.0
xlrd: 0.9.2
xlwt: 0.7.5
xlsxwriter: None
lxml: None
bs4: 4.2.1
html5lib: 0.999
httplib2: 0.8
apiclient: None
sqlalchemy: None
pymysql: 0.7.2.None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None

```

Regards
Stephen

### PR DESCRIPTION
**Pull request by ivannz**
- [x] closes #13813
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by ivannz** [comment]
Upon further consideration I came to the conclusion that the FIX could also have looked like this:
```python
...
                # Skip null keys
                if lab == -1:
                    out[ii] = -1
                    continue
...
```
Now, this raises the question of how groups with partially missing keys should be handled:
* either partial matching of group keys is used, in which case the *-1* in labels should be converted to proper group identifiers (use [get_group_index](https://github.com/pydata/pandas/blob/master/pandas/core/groupby.py#L4062) with ``xnull=False`` in [_get_compressed_labels](https://github.com/pydata/pandas/blob/master/pandas/core/groupby.py#L1649));
* or the incomplete keys are ignored (like in **nunique()**, **value_counts()**), and specifically in **shift()** are filled with missing values.

What do you think, @jreback?

**Comment by jreback** [comment]
pls add some tests

**Comment by ivannz** [comment]
Ok, I will make a concise test and add it to [test_groupby.py](https://github.com/pydata/pandas/blob/master/pandas/tests/test_groupby.py).

**Comment by chris-b1** [comment]
Note that the file you're editing is from a template, you will need to make the change here:
https://github.com/pydata/pandas/blob/master/pandas/src/algos_groupby_helper.pxi.in

I think your second approach is correct (or at least consistent) - in fact the current impl seems to handle it?

```python
In [1]: pd.__version__
Out[1]: u'0.18.1'

In [2]: df = pd.DataFrame([(i%12, i%3 if i%3 else float("nan"), i)
   ...:                    for i in range(17)], dtype=float,
   ...:                   columns=["B", "C", "F"], index=None)

In [3]: df.groupby(['B', 'C']).shift(-1)
Out[3]:
       F
0    NaN
1   13.0
2   14.0
3    NaN
4   16.0
5    NaN
6    NaN
7    NaN
8    NaN
9    NaN
10   NaN
11   NaN
12   NaN
13   NaN
14   NaN
15   NaN
16   NaN
```

File: pandas/src/algos_groupby_helper.pxi [hunk 1]
```diff
@@ -1356,6 +1356,12 @@ def group_shift_indexer(int64_t[:] out, int64_t[:] labels,
                 ## reverse iterator if shifting backwards
                 ii = offset + sign * i
                 lab = labels[ii]
+
```
Comment (sinhrks) [inline pandas/src/algos_groupby_helper.pxi]: do not modify ``pxi`` directly. It is generated from corresponding ``pxi.in``.

File: pandas/tests/test_groupby.py [hunk 2]
```diff
@@ -6560,6 +6560,31 @@ def test_grouping_string_repr(self):
...
 
+    def test_group_shift_with_null_key(self):
+        # This test is designed to replicate the segfault in issue #13813.
+        n_rows = 1200
+
+        # Generate a moderately large dataframe with occasional missing
+        # values in column `B`, and then group by [`A`, `B`]. This should
+        # force `-1` in `labels` array of `gr_.grouper.group_info` exactly
+        # at those places, where the group-by key is partilly missing.
+        df = pd.DataFrame([(i % 12, i % 3 if i % 3 else float("nan"), i)
```
Comment (jreback) [inline pandas/tests/test_groupby.py]: use ``np.nan``

**Comment by ivannz** [comment]
Can somebody tell me what my actions should be when the Travis CI job [#21159](https://travis-ci.org/pydata/pandas/builds/148239405) of this PR fails, whilst the [build job](https://travis-ci.org/ivannz/pandas/builds/148239391) for the same commit on my forked (and rebased) repository succeeded.

**Comment by jorisvandenbossche** [comment]
@ivannz Something went wrong with that build, but probably not your fault. I restarted the build

**Comment by ivannz** [comment]
Great! Thank you, @jorisvandenbossche .

**Comment by jreback** [comment]
thanks @ivannz

**Pull request closed without merging**